ArjunaJTA/jta/tests/classes/com/hp/mwtests/ts/jta/recovery/RecoverySuspendTest.java

```diff
@@ -0,0 +1,221 @@
+
+        // Let's go quick on this
+        _recoveryConfig.setRecoveryBackoffPeriod(recoveryBackoffPeriod);
+        _recoveryConfig.setPeriodicRecoveryPeriod(periodicRecoveryPeriod);
+        /*
+         * Set this interval to be equal to periodicRecoveryPeriod.
+         * The reason behind this choice is:
+         * 1) Activate the Recovery Manager
+         * 2) Create an XA transaction
+         * 3) After periodicRecoveryPeriod seconds, the XA txn can be
+         * processed by the XA Recovery Module
+         */
+        _jtaEnvironmentBean.setOrphanSafetyInterval(periodicRecoveryPeriod);
+        // don't sign off until the store is empty
+        _recoveryConfig.setWaitForWorkLeftToDo(true);
+
+        _recoveryConfig.setRecoveryModuleClassNames(modules);
+
+        _jtaEnvironmentBean.setXaResourceRecoveryClassNames(
+                List.of(BytemanControlledXAResourceRecovery.class.getName()));
+
+        jtaPropertyManager.getJTAEnvironmentBean().setXaRecoveryNodes(List.of("*"));
+    }
+
+    @AfterEach
+    public void afterEach() {
+        _manager.terminate();
+
+        // Let's clean
+        ThreadActionData.purgeActions();
+
+        if (!TxControl.isEnabled()) {
+            // Re-activate transaction creation
+            TxControl.enable();
+        }
+    }
+
+    @Test
+    @BMScripts(scripts = {
+            @BMScript("RecoverySuspendTest/recoverySuspendTest_BytemanControlledXAResource"),
+            @BMScript("RecoverySuspendTest/recoverySusptenTest_BytemanControlledXAResource_commitFailure"),
+            @BMScript("RecoverySuspendTest/recoverySusptenTest_BytemanControlledXAResource_recoverFailure"),
+            @BMScript("RecoverySuspendTest/recoverySuspendTest_FailTest")
+    })
+    public void testSuspensionWhenThereIsXATransactionsToRecover() {
+
+        // Make sure that the test environment is ready
+        resetCounters();
+        BytemanControlledXAResource.setCommitReturn(XAException.XAER_RMFAIL);
+
+        /*
+         * BytemanControlledXAResource.getCommitCallCounter() should be 2 as:
+         * - One invocation from the normal commit procedure should fail
+         * - One invocation from the recovery process should pass
+         * (i.e. when processing AtomicActionRecoveryModule)
+         * BytemanControlledXAResource.getRecoverCallCounter() should be 1 as:
+         * - One invocation comes from the recovery cycle that recovers the transaction
+         */
+        startTest(2, 0, 1, 0);
+    }
+
+    @Test
+    @BMScripts(scripts = {
+            @BMScript("RecoverySuspendTest/recoverySuspendTest_BytemanControlledXAResource"),
+            @BMScript("RecoverySuspendTest/recoverySusptenTest_BytemanControlledXAResource_commitFailure"),
+            @BMScript("RecoverySuspendTest/recoverySusptenTest_BytemanControlledXAResource_rollbackFailure"),
+            @BMScript("RecoverySuspendTest/recoverySuspendTest_SkipAtomicActionRecoveryModule"),
```

Comment:
Please can I ask if this part is strictly necessary? If it is I would be curious to know why and whether not doing AARM functionality might hide a possible problem in the interactions between XARM and AARM? If it's not, I presume it is just some optimization to speed up the test somehow?

Reply:
AARM is used only at the end of the test to clean the environment. IOW, instead of cleaning the transaction myself, I let the Recovery Manager (through AARM) commit the transaction (so that the Recovery manager won't get stuck during suspension). DISCLAIMER: this is an artificial scenario as the transaction that the AARM eventually commits has been already rolled back by the XA resource (locally to itself)...so that transaction should really end up with an heuristic outcome :face_with_peeking_eye: Thus, the outcome I "forced" through the use of AARM isn't specs compliant but it is good enough to clean the environment and test the XARM in isolation.

Reply:
Got it - thank you! I think you should add a comment in this bit of the code too (and in the rule if you didn't already). I like where https://github.com/jbosstm/narayana/pull/2336/commits/15ed2341ecaad32336d2ded0a4819578c8550ea8 is going but I did add a comment there with a thought.

Reply:
Comment added to the test. Thanks for the suggestion :-)